bigtable-client-core-parent/bigtable-hbase/src/main/java/com/google/cloud/bigtable/hbase/wrappers/veneer/DataClientVeneerApi.java

```diff
@@ -230,41 +246,67 @@ protected void onCompleteImpl() {
 
   /** wraps {@link ServerStream} onto HBase {@link ResultScanner}. */
   private static class RowResultScanner extends AbstractClientScanner {
+    // Percentage of max number of rows allowed in the buffer
+    private static final double WATERMARK_PERCENTAGE = .1;
+    private static final RowResultAdapter RESULT_ADAPTER = new RowResultAdapter();
 
     private final Meter scannerResultMeter =
         BigtableClientMetrics.meter(BigtableClientMetrics.MetricLevel.Info, "scanner.results");
     private final Timer scannerResultTimer =
         BigtableClientMetrics.timer(
             BigtableClientMetrics.MetricLevel.Debug, "scanner.results.latency");
 
-    private final ServerStream<Result> serverStream;
-    private final Iterator<Result> iterator;
-
-    RowResultScanner(ServerStream<Result> serverStream) {
-      this.serverStream = serverStream;
-      this.iterator = serverStream.iterator();
+    private ServerStream<Result> serverStream;
+    private ByteString lastSeenRowKey = ByteString.EMPTY;
+    private final Queue<Result> buffer;
+    private final Query.QueryPaginator paginator;
+    private final paginatorFunction wrapper;
+    private final int refillSegmentWaterMark;
+
+    RowResultScanner(Query.QueryPaginator paginator, paginatorFunction wrapper) {
+      this.paginator = paginator;
+      this.wrapper = wrapper;
+      this.buffer = new ArrayDeque<>();
+      this.refillSegmentWaterMark = (int) (PAGE_SIZE * WATERMARK_PERCENTAGE);
```

Comment:
You need to do Math.max(1, (int) (PAGE_SIZE * WATERMARK_PERCENTAGE) to avoid not reading all the data when the page size is small.

Reply:
In the beam client it made sense. Here I am multiplying two private static final variables :)